### PR TITLE
Launch control refactor (up to date with competition VCU code)

### PR DESCRIPTION
**Pull request by sebLopezCot**

### CODE REVIEW

File: src/vcu/inc/controls.h [hunk 1]
```diff
@@ -14,6 +14,15 @@
 #define MAX_ACCEL_VAL 1000
 #define MIN_ACCEL_VAL 0
 
+// Launch control contants
```
Comment (sebLopezCot) [inline src/vcu/inc/controls.h:17]: spelling error

File: src/vcu/inc/controls.h [hunk 2]
```diff
@@ -14,6 +14,15 @@
...
+// After wheel speed crosses this threshold, start using slip controller
+#define LC_WS_THRESH 45000
+#define LC_cGR 347 // Gear ratio times 100
+#define LC_ACCEL_BEGIN 950 // 95%
+#define LC_ACCEL_RELEASE 50 // 5%
+#define LC_BRAKE_BEGIN 100 // We want a lower threshold
```
Comment (sebLopezCot) [inline src/vcu/inc/controls.h:23]: Do we still want this comment?

File: src/vcu/src/can_handles.c [hunk 3]
```diff
@@ -293,6 +293,17 @@ void handleDashRequest(Frame *msg) {
   control_settings.active_aero_enabled = unpacked_msg.active_aero_enabled;
 }
 
+void handle_DashRequestLC(Frame *msg) {
+  can0_DashRequestLC_T unpacked_msg;
+
+  unpack_can0_DashRequestLC(msg, &unpacked_msg);
+
+  if (unpacked_msg.launch_ctrl_slip_ratio != 65535) {
```
Comment (sebLopezCot) [inline src/vcu/src/can_handles.c]: What is 65535? Should this really go in the Dash Request handler?

File: src/vcu/src/controls.c [hunk 4]
```diff
@@ -69,9 +80,10 @@ void execute_controls(void) {
   // Extra check to ensure we are only sending regen torque when allowed
   if (torque_command == 0) {
     torque_command = regen_torque;
-  }
-  else {
-    // Only use limits when we're not doing regen
+  } else { // Only use limits and launch control when we're not doing regen
+    // Even if we are doing launch control, calculate limits for loggind also so that we are not doing
```
Comment (sebLopezCot) [inline src/vcu/src/controls.c]: spelling error "loggind"

File: src/vcu/src/controls.c [hunk 5]
```diff
@@ -69,9 +80,10 @@ void execute_controls(void) {
...
-  }
-  else {
-    // Only use limits when we're not doing regen
+  } else { // Only use limits and launch control when we're not doing regen
+    // Even if we are doing launch control, calculate limits for loggind also so that we are not doing 
+    // launch control when we're supposed to be limiting
+
     int32_t voltage_limited_torque = get_voltage_limited_torque(torque_command);
     // static uint32_t last_vt = 0;
     // if (HAL_GetTick() - last_vt > 10) {
```
Comment (sebLopezCot) [inline src/vcu/src/controls.c:90]: Just remove this commented timer code.

File: src/vcu/src/controls.c [hunk 6]
```diff
@@ -47,6 +57,7 @@ bool get_controls_enabled(void) {
 
 void execute_controls(void) {
```
Comment (sebLopezCot) [inline src/vcu/src/controls.c:59]: Not paramount, but this function is really long and complicated. Can it be broken up into smaller functions? Also can the logic be structured to use fewer nested if / switch statements / conditionals (&&, !, ||)?

File: src/vcu/src/controls.c [hunk 7]
```diff
@@ -105,15 +118,72 @@ void execute_controls(void) {
...
+          sendSpeedCmdMsg(0, 0);
+
+          // Transition
+          if (pedalbox_min(accel) > LC_ACCEL_BEGIN) {
+            lc_state = SPEEDING_UP;
+            printf("[LAUNCH CONTROL] SPEEDING UP STATE ENTERED\r\n");
+          }
+          break;
+        case SPEEDING_UP:
+          if (1000 > torque_command) sendSpeedCmdMsg(500, torque_command);
```
Comment (sebLopezCot) [inline src/vcu/src/controls.c]: Why 1000? What does this variable indicate? Why 500? What does this variable indicate?

File: src/vcu/src/controls.c [hunk 8]
```diff
@@ -105,15 +118,72 @@ void execute_controls(void) {
...
+          else sendSpeedCmdMsg(300, 1000);
```
Comment (sebLopezCot) [inline src/vcu/src/controls.c]: Same here

File: can_spec_my18.yml [hunk 9]
```diff
@@ -681,6 +671,19 @@ buses:
             position: 60
             length: 1
             c_type: bool
+      DashRequestLC:
+        can_id: 0x0D5
+        is_big_endian: true
+        period: 69ms
```
Comment (sebLopezCot) [inline can_spec_my18.yml:682]: Lol 69

File: can_spec_my18.yml [hunk 10]
```diff
@@ -681,6 +671,19 @@ buses:
...
+        segments:
+          using_launch_ctrl:
+            position: 0
+            length: 1
+            c_type: bool
+          launch_ctrl_slip_ratio: # Ranges from 105 to 115
```
Comment (sebLopezCot) [inline can_spec_my18.yml:688]: Why does it range from 105 to 115?

File: can_spec_my18.yml [hunk 11]
```diff
@@ -681,6 +671,19 @@ buses:
...
+            position: 16
+            length: 16
+            c_type: uint16_t
```
Comment (sebLopezCot) [inline can_spec_my18.yml:691]: Could we use a uint8_t instead since it only ranges between 105 and 115?

File: can_spec_my18.yml [hunk 12]
```diff
@@ -736,6 +739,19 @@ buses:
...
+        period: 200ms
+        segments:
+          using_launch_ctrl:
+            position: 0
+            length: 1
+            c_type: bool
+          slip_ratio: # Ranges from 105 to 115
+            position: 16
+            length: 16
+            c_type: uint16_t
```
Comment (sebLopezCot) [inline can_spec_my18.yml:753]: Same comments here as above

File: src/dashboard/src/page_manager.c [hunk 13]
```diff
@@ -430,6 +434,34 @@ void draw_volt_lim_page(page_manager_t *pm, NHD_US2066_OLED *oled) {
...
+    carstats_t *stats = pm->stats;
+
+    oled_clearline(oled, 1);
+    oled_set_pos(oled, 1, 0);
+
+    if (stats->buttons.right.rising_edge) {
+        stats->controls.launch_ctrl_slip_ratio += 1;
+    }
+
+    stats->controls.launch_ctrl_slip_ratio = LOOPOVER(stats->controls.launch_ctrl_slip_ratio, 101, 120);
```
Comment (sebLopezCot) [inline src/dashboard/src/page_manager.c:447]: What does LOOPOVER do? Is it a custom construct?

File: src/vcu/src/controls.c [hunk 14]
```diff
@@ -22,6 +28,9 @@ void init_controls_defaults(void) {
   control_settings.volt_lim_min_gain = 0;
   control_settings.volt_lim_min_voltage = 300;
   control_settings.torque_temp_limited = false;
+
+  lc_settings.using_launch_ctrl = false;
+  lc_settings.slip_ratio = 112;
```
Comment (sebLopezCot) [inline src/vcu/src/controls.c:33]: Might want to bring 112 out to a DEFAULT_SLIP_RATIO constant

**Review by sebLopezCot — COMMENTED** [review]
An initial look at the PR. Most of this is stylistic. I can make these changes if given the okay. I still want to do a more in depth analysis of how the algorithm works.

File: src/vcu/src/can_handles.c (hunk 3 above)
Comment (nistath) [reply src/vcu/src/can_handles.c]: -1 is the convention for "hey vcu use your default value". Should've been -1 cast to the correct type.

File: can_spec_my18.yml (hunk 11 above)
Comment (nistath) [reply can_spec_my18.yml:691]: Premature opt.

File: src/dashboard/src/page_manager.c (hunk 13 above)
Comment (nistath) [reply src/dashboard/src/page_manager.c:447]: It is. If the value goes above the third argument then it loops over to the second argument using modulo.

**Comment by nistath** [comment]
Closed to avoid confusion. Please move all review comments to #23 .

**Pull request closed without merging**